### ISSUE TITLE
Institution/Ministry - Deactivate Program - UI Changes

### ISSUE THREAD

**Issue by HRAGANBC** [bcgov/SIMS#2645]
### **User Story**
As an _Institution or Ministry User_, I may need to deactivate a program. This could be due to no longer offering it, or creating it in error resulting in a new revised program being created.

### **Acceptance Criteria**
- [x] Add new state for programs: Inactive (preserve current status at time of deactivation)
- [x] If the state is inactive, update status flag for programs to show the "Inactive" flag in dashboards. 
- [x] Ministry User: Update edit button on Program details screen to drop down with two options: Edit Program, Deactivate. (No change to functionality of "Edit Program")
- [x] Make this feature/role available to **all** Groups within Ministry KeyCloak configs
- [x] Institution User: Update edit button on Program details screen to drop down with two options: Edit Program, Deactivate. (No change to functionality of "Edit Program")
- [x] New pop up when user selects "Deactivate" from edit program details button
- [x] When the Ministry is deactivating, include a basic `Notes` section (not part of the mockup) to save a note to the institution's notes under the `Program` category.
- [x] Update new program creation validation to ignore any deactivated programs. (i.e. so that the same SABC Program Code can be re-used on the new Programs)
- [x] Update bulk upload to ignore deactivated programs when assigning offerings.    

**Technical**

- [x] Create a new column `sims.education_programs.is_active` to represent the inactive state of the program.
- [x] Create new specific audit columns to save the user deactivating and the date, for instance, `deactivated_by`, and `deactivation_date` (we may create columns also as `is_active_updated_on` and `is_active_updated_by`).
- [x] Adjust the badge status component on the web portal to display `Inactive` when the `is_active` is false, otherwise, display the status as it is.
- [x] Create a new role for the Ministry approval action. The new role should be associated with all groups as per stated in the business AC.
- [x] Create API E2E tests for the new endpoints.

### **Mock ups**
**Update to "Edit" button on Program Details to a Drop-down selection with new pop-up modal:**
<img width="682" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBME10QXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--70ade47a3a4d48385754b84b3d45b085fb45f971/image.png" alt="image.png" />      
### **Current Behaviour**
Programs exist _forever_ and institution users just update the program name/description to indicate it should not be used.
Programs with the same SABC program code cannot exist at the same time.

### **Expected Behaviour**
Institution users AND Ministry users can "deactivate" a program:
- This will change the flag beside each program to indicate it has a different status. <img width="141" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNmdsQXc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--19dce0c6e55994809a278050cac72ab3db241450/image.png" alt="image.png" />
- Programs with this status will no longer be visible to students or PSIs in dropdowns for student applications/PIRs. (Applications, PIR)

**Pending Programs:**
- Programs that are Pending can be deactivated.

**Options once deactivated:**
- Deactivated programs can not be "reactivated" by PSIs. (PSIs can make a new program)
- Deactivated programs can not have new offerings added to it: Disable UI options & bulk upload will ignore them if SABC program codes match
- This will not prevent any system created offerings/duplicates to support scholastic standing,etc.
- FUTURE POSSIBLE ENHANCEMENT: Deactivated programs can be "reactivated" by Ministry with an additional prompt to confirm that the program meets all current requirements.

**Comment by HRAGANBC** [comment]
@CarlyCotton - really like this ticket, I think it will be great improvement. 

- Likely for the Devs to sort out during implementation, but are we sure changing the status to 'inactive' doesn't muck up the student's NoA, the IER, others back end stuff?
- Can users deactive a program when it is still in 'pending' status?
- What happens to attached students who need to change study information mid-application? Like they need to change study period info which is initiated by the school assigning them to new programs & offerings via dropdown menus? I think we're fine with business process where they'll have to assign them to entirely new program/offering, right?  
- What happens to offerings which are cloned during scholastic standing - they should still be added to the deconmissioned program. We just basically disable the UX/UI button for 'add offering' as well as the Bulk Upload route. 
- Should programs with effective dates which 'expire' automaticaly change to 'inactive' status? What happens to them currently when they reach their effective end date (I believe they stop appearing in student dropdowns, but still appear as active in th Institution's program inventory.


Some other things for us to consider that might need seperate tickets:
- This doesn't solve the issue of "D restrictions", i.e. how to shut down a program to new entrants while letting existing affiliated students continue to submit **new, future** applicaitons to that program. 
- We'll likely need an equivalent way of decomissioning offerings, but let's learn from this ticket first and see if any tweaks are needed

**Comment by CarlyCotton** [comment]
>Likely for the Devs to sort out during implementation, but are we sure changing the status to 'inactive' doesn't muck up the student's NoA, the IER, others back end stuff?

Interested to hear more from them re impacts on existing applications or anything else.

>Can users deactive a program when it is still in 'pending' status?

I think they should - this would be helpful if they made one that they didn't expect to be pending. It'll save time for ministry users and institutions for them to just deactivate.

>What happens to attached students who need to change study information mid-application? Like they need to change study period info which is initiated by the school assigning them to new programs & offerings via dropdown menus? I think we're fine with business process where they'll have to assign them to entirely new program/offering, right?

If it's in draft, I expect it to handle like other changes to designation/offerings and just removes the drop down selection and forces a new selection. I'll clarify this in the expected behaviour.

> What happens to offerings which are cloned during scholastic standing - they should still be added to the deconmissioned program. We just basically disable the UX/UI button for 'add offering' as well as the Bulk Upload route.

I think this is a good distinction - not blocking the add offering, but disabling it from a UI standpoint and ignoring them during bulk upload. 

>Should programs with effective dates which 'expire' automaticaly change to 'inactive' status? What happens to them currently when they reach their effective end date (I believe they stop appearing in student dropdowns, but still appear as active in th Institution's program inventory.

They currently continue to show a status of "Approved" and are not visible in new application drop-downs. They are not able to have the expiry extended through another review process, but right now if they had a SABC program code to reuse they wouldn't be able to re-create.


>We'll likely need an equivalent way of decomissioning offerings, but let's learn from this ticket first and see if any tweaks are needed

This is something I considered. Offerings can be "decommissioned" by changing it to private and they eventually drop off based on PY cycle. The concept of private programs could start resolving D restrictions.

**Comment by andrewsignori-aot** [comment]
@CarlyCotton questions from DEVs:
- Should we capture the user deactivating the program (Institution or Ministry)?
- If the Ministry is deactivating it, should we add a note to the Institution notes?

**Comment by CarlyCotton** [comment]
@andrewsignori-aot 
1. Please keep a record of who made the change, this doesn't need to be displayed on the Program.
2. Yes please! If Ministry users could be asked to add an Institution Note with a type of "Program" when doing the deactivate. (if this is a large functionality add please let me know and we can consider it for future enhancement)

**Comment by andrewsignori-aot** [comment]
@CarlyCotton 
Regarding number 1, I believe that makes sense too. We did it in other areas.
Regarding number 2, we have some framework in place and by default every time that Ministry does an action "on behalf" of the institution we try to insert a related note.
I will add these two items to the technical details.
Thanks for the clarifications 😉

**Comment by andrewsignori-aot** [comment]
@CarlyCotton 
I added item 1 as a technical AC and item 2 as a business AC: "When the Ministry is deactivating, include a basic Notes section (not part of the mockup) to save a note to the institution's notes under the Program category."

**Comment by andrewsignori-aot** [comment]
@CarlyCotton regarding the Ministry approving an offering-related change request (Application Offering Change and Bulk Offering change) should the Ministry be allowed to approve a request for an inactive program?
Should the student be able to approve an Application Offering Change for an inactive program?

**Comment by andrewsignori-aot** [comment]
@CarlyCotton for implementation we are splitting this ticket into two. The first ticket will change in the UI changes and the second one in applying the constraints (#2930).

**Comment by CarlyCotton** [comment]
@andrewsignori-aot 
> regarding the Ministry approving an offering-related change request (Application Offering Change and Bulk Offering change) should the Ministry be allowed to approve a request for an inactive program?
> Should the student be able to approve an Application Offering Change for an inactive program?

A change to an offering itself **should** be able to be actioned regardless of the program being active/inactive. There needs to be a way for these to be actioned out of the pending offering queue and we don't want to auto-decline simply because the program is inactive. (It is possible that a change needs to be made on the offering with applications attached, which we want to keep functioning as normal. Just no NEW applications to these. (Possibility of future re-activation enhancements would also be affected by auto-declining.)

**Request A Change process**
At this time we don't want the schools to be able to select the deactivated program from the Request A Change feature. 
If there are pending requests (waiting on student or ministry action) to change an application into a program that becomes deactivated, we don't want to stop that process. If it was already "in motion" before the deactivation, it can continue "in motion".

**Comment by andrewsignori-aot** [comment]
> @andrewsignori-aot
> 
> > regarding the Ministry approving an offering-related change request (Application Offering Change and Bulk Offering change) should the Ministry be allowed to approve a request for an inactive program?
> > Should the student be able to approve an Application Offering Change for an inactive program?
> 
> A change to an offering itself **should** be able to be actioned regardless of the program being active/inactive. There needs to be a way for these to be actioned out of the pending offering queue and we don't want to auto-decline simply because the program is inactive. (It is possible that a change needs to be made on the offering with applications attached, which we want to keep functioning as normal. Just no NEW applications to these. (Possibility of future re-activation enhancements would also be affected by auto-declining.)
> 
> **Request A Change process** At this time we don't want the schools to be able to select the deactivated program from the Request A Change feature. If there are pending requests (waiting on student or ministry action) to change an application into a program that becomes deactivated, we don't want to stop that process. If it was already "in motion" before the deactivation, it can continue "in motion".

Thanks for the clarification @CarlyCotton.

Regarding editing an offering we have:
- Scenario A, "no application is associated with the offering", the institution can freely edit the offering and programs are not involved in this.
- Scenario B, "some application is associated with the offering" which will create new "versions" of the same offering for history and for the Ministry approval. For this scenario we should allow the system to create the new offerings under the program even if it is deactivated, right?

**Comment by CarlyCotton** [comment]
@andrewsignori-aot 

> Scenario A, "no application is associated with the offering", the institution can freely edit the offering and programs are not involved in this.
> Scenario B, "some application is associated with the offering" which will create new "versions" of the same offering for history and for the Ministry approval. For this scenario we should allow the system to create the new offerings under the program even if it is deactivated, right?

Scenario A - exactly
Scenario B - it's fine for biz if the system creates a new offering on a deactivated program.

**Comment by andrewsignori-aot** [comment]
@HRAGANBC checking Keycloak, we have been creating the new roles just under admins groups. Related to the below AC
"Make this feature/role available to all Groups within Ministry KeyCloak configs"
can we create just under the `aest-business-administrators` till we have further clarifications on the roles?
@CarlyCotton @Joshua-Lakusta FYI

**Comment by andrewsignori-aot** [comment]
@CarlyCotton there is a minor difference between Institution and Ministry program actions.
![image](https://github.com/bcgov/SIMS/assets/61259237/e854251e-95a0-411a-bace-2cbe34375c3f)

The institution will display "Edit program" while the Ministry will see "View program".
When program are no longer editable, for instance, when an offering is added, the menus also display "View program".
Can we have the action button named as "Program actions" for both the Ministry and institutions, as shown below?
Also I will disable the "Deactivate" button and the program page will be displayed as read-only. Does it make sense?

![image](https://github.com/bcgov/SIMS/assets/61259237/7a419e84-621e-48d8-8172-b7abc8d090b4)

![image](https://github.com/bcgov/SIMS/assets/61259237/f64505bf-c045-4061-a533-b62f504d836c)

**Comment by CarlyCotton** [comment]
@andrewsignori-aot The changes look great. Business is good with the Deactivate being blocked out and the switch from "Edit program" to "View Program" once deactivated.

**Comment by andrewsignori-aot** [comment]
As agreed with @CarlyCotton even though not explicitly mentioned in the AC, the program view for inactive programs is read-only for now for consistency and to avoid further updated in a no longer active program.
